edgelet/edgelet-image-cleanup/src/image_gc.rs

```diff
@@ -0,0 +1,205 @@
+
+    let bootstrap_img_id = match bootstrap_image_id_option.clone() {
+        Some(id) => id,
+        None => String::default(),
+    };
+
+    // track images associated with extant containers
+    let modules = ModuleRuntime::list_with_details(runtime)
+        .await
+        .map_err(ImageCleanupError::ListRunningModules)?;
+    let mut in_use_image_ids: HashSet<String> = HashSet::new();
+    if !bootstrap_img_id.is_empty() {
+        // the bootstrap edge agent image should never be deleted.
+        in_use_image_ids.insert(bootstrap_img_id.clone());
+    }
+
+    for module in modules {
+        // this is effectively the result of calling GET on /containers/json
+        // image_hash() is created from ImageId (ImageId is filled in by GET /containers/json)
+        let id = edgelet_core::Module::config(&module.0)
+            .image_hash()
+            .ok_or(ImageCleanupError::GetImageId())?;
+        in_use_image_ids.insert(id.to_string());
+    }
+
+    let image_map = image_use_data
+        .prune_images_from_file(in_use_image_ids)
+        .map_err(ImageCleanupError::PruneImages)?;
+
+    // delete images
+    for key in image_map.keys() {
+        if let Err(e) = ModuleRegistry::remove(runtime, key).await {
+            log::error!("Could not delete image {} : {}", key, e);
+        }
+    }
+
+    Ok(())
+}
+
+/* ================================================ HELPER METHODS ================================================ */
+
+// This is a helper method that gets the imageID of the bootstrap edge agent, if it is present on the box.
+// This image is used as a fallback in certain scenarios and we have to make sure that we never delete it
+// (though the customer can still do so).
+//
+// To wit, there are 4 possibilities (pertaining to this method's return values):
+//    Image Id found       Has image been deleted?     Interpretation
+//        yes                        yes               N/A [can't happen]
+//        yes                        no                prune images, as per usual
+//        no                         yes               customer deleted bootstrap; prune images, as per usual
+//        no                         no                (Implicit) Docker Engine API error; update persistence file but
+//                                                     do not prune images to ensure EA bootstrap isn't deleted
+
+async fn get_bootstrap_image_id(
+    runtime: &edgelet_docker::DockerModuleRuntime<http_common::Connector>,
+    edge_agent_bootstrap: String,
+) -> Result<(String, bool), ImageCleanupError> {
+    let image_name_to_id = ModuleRuntime::list_images(runtime)
+        .await
+        .map_err(ImageCleanupError::ListImages)?;
+    let image_id_option = image_name_to_id.get(&edge_agent_bootstrap);
+
+    let bootstrap_image_id =
+        image_id_option.map_or_else(String::default, |image_id| image_id.to_string());
+
+    Ok((bootstrap_image_id.clone(), bootstrap_image_id.is_empty()))
```

Comment:
Here you convert an `Option<String>` to a tuple `(String, bool)`, which the caller converts back to an `Option<String>`. Why not just return `Result<Option<String>>`? Are you doing this because `list_images()` might return a HashMap entry with an empty value?